server/src/main/scala/org/http4s/server/middleware/BracketRequestResponse.scala

```diff
@@ -0,0 +1,197 @@
+/*
+ * Copyright 2013-2020 http4s.org
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.http4s.server.middleware
+
+import cats.data._
+import cats.effect._
+import cats.effect.syntax.all._
+import cats.implicits._
+import org.http4s._
+import org.http4s.server._
+
+/** Middelwares which allow for bracketing on a Request/Response, including
+  * the completion of the Response body stream.
+  *
+  * These are analogous to `cats.effect.Bracket` and `fs2.Stream.bracket`. The
+  * reason that they exist is because due to the full termination of a
+  * Response being a function of the termination of the `fs2.Stream` which
+  * backs the response body, you can't actually use either
+  * `cats.effect.Bracket` or `fs2.Stream.bracket` directly.
+  *
+  * http4s has encodings similar to these in the main repo, but they are all
+  * special cased. Having a generic implementation opens up the possibility of
+  * writing many interesting middlewares with ease.
+  *
+  * @define releaseWarning The bracketing semantics defined here differ in one
+  *         very important way from cats-effect or fs2 bracketing
+  *         semantics. If the [[Response]] body is not evaluated after the
+  *         application of the middleware, then the `release` handler ''will
+  *         not run, effectively creating a resource leak.'' This can happen
+  *         due to a bug in a server implementation, where it fails to drain
+  *         the body, or due to a user code if the [[Response]] body stream
+  *         (or the [[Response]] itself) is discarded after the application of
+  *         this middleware. Unfortunately, there is currently no way to avoid
+  *         this. For this reason, it is strongly recommended that this
+  *         middleware be applied at the most outer scope possible. Appending
+  *         to or transforming the [[Response]] body ''is safe'' to do after
+  *         the application of this middleware.
+  */
+object BracketRequestResponse {
+
+  /** A [[Response]] value with some arbitrary context added. */
+  final case class ContextResponse[F[_], A](context: A, response: Response[F])
```

Comment:
I think there are still bigger thoughts to be had about this and `ContextRequest` and the old idea of turning message into a typeclass like @ChristopherDavenport has been proposing forever, but that's a 1.0 discussion.